### PR TITLE
Feature: Multi Target Update Support

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR is an enhancement on earlier PR to add support for multi-target update.

Example: `update user u join m on u.col = m.col set u.foo = 32, m.bar = 23 where u.id = 2`
The SQL here updates columns from two tables using the join condition.

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/15211

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: changelog/20.0/20.0.0/summary.md [hunk 1]
```diff
@@ -9,7 +9,9 @@
     - [Vindex Hints](#vindex-hints)
     - [Update with Limit Support](#update-limit)
     - [Update with Multi Table Support](#multi-table-update)
+    - [Update with Multi Target Support](#update-multi-target)
```
Comment (frouioui) [inline changelog/20.0/20.0.0/summary.md:12]: Query compatibility is going to be a great theme for v20! ❤️

File: go/vt/vtgate/planbuilder/operators/update.go [hunk 2]
```diff
@@ -926,3 +998,113 @@ func nullSafeNotInComparison(ctx *plancontext.PlanningContext, updatedTable *vin
...
+		return nil, nil, nil
+	}
+	// generate rest of the owned vindex query.
+	ovq := &sqlparser.Select{
+		SelectExprs: selExprs,
+		OrderBy:     update.OrderBy,
+		Limit:       update.Limit,
+		Lock:        sqlparser.ForUpdateLock,
+	}
+	return changedVindexes, ovq, subQueriesArgOnChangedVindex
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/update.go:1047]: it it not super important but it is a bit confusing why we are naming the return arguments on line `1008` but it seems like there is no reason to not directly used the variable declared in `1008`.

**Review by frouioui — APPROVED** [review]

File: go/test/endtoend/vtgate/queries/dml/dml_test.go [hunk 3]
```diff
@@ -368,3 +368,40 @@ func TestMultiTargetDeleteMore(t *testing.T) {
...
+	utils.SkipIfBinaryIsBelowVersion(t, 20, "vtgate")
+
+	mcmp, closer := start(t)
+	defer closer()
+
+	// initial rows
+	mcmp.Exec("insert into order_tbl(region_id, oid, cust_no) values (1,1,4), (1,2,2), (2,3,5), (2,4,55)")
+	mcmp.Exec("insert into oevent_tbl(oid, ename) values (1,'a'), (2,'b'), (3,'a'), (4,'c')")
+
+	// check rows
```
Comment (systay) [inline go/test/endtoend/vtgate/queries/dml/dml_test.go:383]: nitpick: do we really need this check?

File: go/vt/sqlparser/ast_funcs.go [hunk 4]
```diff
@@ -2686,3 +2686,10 @@ func (node *Update) SetWherePredicate(expr Expr) {
 		Expr: expr,
 	}
 }
+
+func (lock Lock) Merge(newLock Lock) Lock {
```
Comment (systay) [inline go/vt/sqlparser/ast_funcs.go:2690]: not self-evident what this is doing. comment?

File: go/vt/vtgate/planbuilder/operators/update.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/update.go:1045]: fixed

File: go/vt/vtgate/planbuilder/plan_test.go [hunk 5]
```diff
@@ -488,6 +488,7 @@ func TestBypassPlanningKeyrangeTargetFromFile(t *testing.T) {
 }
 
 func TestWithDefaultKeyspaceFromFile(t *testing.T) {
+	t.Skip()
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/plan_test.go:491]: removed

File: go/test/endtoend/vtgate/queries/dml/dml_test.go (hunk 3 above)
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/queries/dml/dml_test.go:383]: it is not interesting, just helps to update the expectation later. removed now.

File: go/vt/sqlparser/ast_funcs.go (hunk 4 above)
Comment (harshit-gangal) [reply go/vt/sqlparser/ast_funcs.go:2690]: renamed the method and added comment.

File: go/vt/vtgate/planbuilder/operators/dml_planning.go [hunk 6]
```diff
@@ -41,6 +39,35 @@ type TargetTable struct {
 	Name   sqlparser.TableName
 }
 
+// dmlOp stores intermediary value for Update/Delete Operator with the vindexes.Table for ordering.
+type dmlOp struct {
+	op   Operator
+	vTbl *vindexes.Table
+	cols []*sqlparser.ColName
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/dml_planning.go:46]: what are we storing in `cols`?

File: go/vt/vtgate/planbuilder/operators/update.go [hunk 7]
```diff
@@ -123,10 +127,54 @@ func createOperatorFromUpdate(ctx *plancontext.PlanningContext, updStmt *sqlpars
...
+}
+
+func isMultiTargetUpdate(ctx *plancontext.PlanningContext, childFks []vindexes.ChildFKInfo, parentFks []vindexes.ParentFKInfo, updateStmt *sqlparser.Update) bool {
+	var targetTS semantics.TableSet
+	for _, ue := range updateStmt.Exprs {
+		targetTS = targetTS.Merge(ctx.SemTable.DirectDeps(ue.Name))
+	}
+	if targetTS.NumberOfTables() == 1 {
+		return false
+	}
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/update.go:157]:
```suggestion
	if ctx.SemTable.Targets.NumberOfTables() == 1 {
		return false
	}
```

File: go/vt/vtgate/planbuilder/update.go [hunk 8]
```diff
@@ -51,7 +50,7 @@ func gen4UpdateStmtPlanner(
 	if ctx.SemTable.HasNonLiteralForeignKeyUpdate(updStmt.Exprs) {
 		// Since we are running the query with foreign key checks off, we have to verify all the foreign keys validity on vtgate.
 		ctx.VerifyAllFKs = true
-		updStmt.SetComments(updStmt.GetParsedComments().SetMySQLSetVarValue(sysvars.ForeignKeyChecks, "OFF"))
+		// updStmt.SetComments(updStmt.GetParsedComments().SetMySQLSetVarValue(sysvars.ForeignKeyChecks, "OFF"))
```
Comment (systay) [inline go/vt/vtgate/planbuilder/update.go:53]: remove?

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/dml_planning.go (hunk 6 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/dml_planning.go:46]: They store the select columns to get from the input.

File: go/vt/vtgate/planbuilder/operators/update.go (hunk 7 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/update.go:157]: This cannot be used, as the ctx remains same for all the update queries generated afterwards, so this condition will either remain true or false for all the newer update queries.

**Pull request merged**